src/common-ssh/ssh.c

```diff
@@ -412,6 +428,84 @@ static int guac_common_ssh_authenticate(guac_common_ssh_session* common_session)
 
 }
 
+/**
+ * Verifies if given algorithms are supported by libssh2 and
+ * makes the |prefs| string to use with libssh2_session_method_pref().
+ * 
+ * If the verification fails, all preferred algorithms |algs| are
+ * added to the |prefs| string.
+ *
+ * @param client
+ *     The Guacamole client that is using SSH.
+ * 
+ * @param session
+ *     The session associated with the user to be authenticated.
+ *
+ * @param method_type
+ *      One of the libssh2 Method Type constants for libssh2_session_method_pref().
+ * 
+ * @param algs
+ *      A preferred list of algorithms, for example FIPS_COMPLIANT_CIPHERS.
+ * 
+ * @return
+ *     The |prefs| string to use with libssh2_session_method_pref().
+ */
+static char *get_method_pref_string(guac_client* client, LIBSSH2_SESSION* session,
+        int method_type, const char** algs) {
+
+    /* Initial size of the string. It will be reallocated later if not enough. */
+    int algs_str_size = 1024;
+    char* algs_str = guac_mem_alloc(algs_str_size);
+    
+    algs_str[0] = 0;
+
+    /* Request the list of supported algorithms/cyphers from libssh2. */
+    const char **supported_algs;
+    int supported_algs_count =
+        libssh2_session_supported_algs(session, method_type, &supported_algs);
+
+    /* This is not a show stopper. */
+    if (supported_algs_count <= 0) {
+        guac_client_log(client, GUAC_LOG_WARNING,
+            "Could not verify if preferred algorithms/ciphers are supported by libssh2");
+    }
+
+    int i = 0;
+    while (algs[i]) {
+        bool found = false;
+        /* Check if the algorithm is found in the libssh2 supported list. */
+        for (int j = 0; j < supported_algs_count; j++) {
+             if (strcmp(algs[i], supported_algs[j]) == 0) {
+                found = true;
+                break;
+            }
+        }
+        /* Add the algorithm to the string if found or libssh2_session_supported_algs() failed. */
+        if (found || supported_algs_count <= 0) {
+            bool req_comma = algs_str[0] != 0;
+            /* Size of the algorithm name plus 1 for comma (if required) plus 1 for '\0'. */
+            int req_size = strlen(algs[i]) + (req_comma ? 1 : 0) + 1;
+            if (strlen(algs_str) + req_size > algs_str_size) {
+                /* Either 1024 to not reallocate often or required size if bigger. */
+                algs_str_size += MAX(req_size, 1024);
+                algs_str = guac_mem_realloc(algs_str, algs_str_size);
```

Comment:
This can fail, which will overwrite `algs_str` with `NULL` and leave a dangling block of memory. `guac_mem_realloc_or_die()` is a better choice if always discarding the original pointer.